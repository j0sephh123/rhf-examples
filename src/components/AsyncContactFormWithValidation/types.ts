export type ContactFormFields = { name: string; email: string; message: string };