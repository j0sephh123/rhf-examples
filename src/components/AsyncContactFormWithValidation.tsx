import clsx from "clsx";
import { useForm } from "react-hook-form";

const mockApi = async (contactFormFields: ContactFormFields) => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      if (
        contactFormFields.email &&
        contactFormFields.message &&
        contactFormFields.name
      ) {
        resolve(undefined);
      } else {
        reject("somethig went wrong");
      }
    }, 500);
  });
};

const emailRegex = /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$/;

const errorsMessages = {
  name: {
    required: "Name is required",
    minLength: "Must be at least two characters long.",
  },
  email: {
    required: "Email is required",
    pattern: "Must include an '@' symbol and a dot.",
  },
  message: {
    required: "Message is required",
    minLength: "Must be at least ten characters long.",
  },
};

type ContactFormFields = { name: string; email: string; message: string };

export default function AsyncContactFormWithValidation() {
  const {
    register,
    handleSubmit,
    formState: { errors, isLoading, isSubmitting },
    reset,
  } = useForm<ContactFormFields>({
    mode: "onSubmit",
    reValidateMode: "onChange",
  });

  console.log({ isLoading, isSubmitting });

  const onSubmit = async (contactFormFields: ContactFormFields) => {
    await mockApi(contactFormFields);

    reset();
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className="form-control w-full max-w-xs">
        <label className="label">
          <span className="label-text">What is your name?</span>
        </label>
        <input
          type="text"
          placeholder="Type here"
          className={clsx(
            "input input-bordered w-full max-w-xs",
            errors.name && "input-error"
          )}
          {...register("name", {
            required: {
              value: true,
              message: "name is required",
            },
            minLength: {
              message: errorsMessages.name.minLength,
              value: 2,
            },
          })}
        />
        {errors.name && (
          <p className="text-error text-sm mt-1">{errors.name.message}</p>
        )}
      </div>
      <div className="form-control w-full max-w-xs">
        <label className="label">
          <span className="label-text">What is your email?</span>
        </label>
        <input
          type="text"
          placeholder="Type here"
          className={clsx(
            "input input-bordered w-full max-w-xs",
            errors.email && "input-error"
          )}
          {...register("email", {
            required: {
              value: true,
              message: "Email is required",
            },
            pattern: {
              value: emailRegex,
              message: errorsMessages.email.pattern,
            },
          })}
        />
        {errors.email && (
          <p className="text-error text-sm mt-1">{errors.email.message}</p>
        )}
      </div>
      <div className="form-control">
        <label className="label">
          <span className="label-text">Your message</span>
        </label>
        <textarea
          className={clsx(
            "textarea textarea-bordered h-24",
            errors.message && "textarea-error"
          )}
          placeholder="Bio"
          {...register("message", {
            required: {
              value: true,
              message: "Message is required",
            },
            minLength: {
              value: 10,
              message: errorsMessages.message.minLength,
            },
          })}
        ></textarea>
        {errors.message && (
          <p className="text-error text-sm mt-1">{errors.message.message}</p>
        )}
      </div>
      <button
        type="submit"
        className={clsx(
          "mt-4 btn-neutral btn btn-block",
          isSubmitting && "btn-disabled"
        )}
      >
        {isSubmitting ? "Loading..." : "Submit"}
      </button>
    </form>
  );
}
