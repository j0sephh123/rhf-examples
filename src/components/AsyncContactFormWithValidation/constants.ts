export const errorsMessages = {
  name: {
    required: "Name is required",
    minLength: "Must be at least two characters long.",
  },
  email: {
    required: "Email is required",
    pattern: "Must include an '@' symbol and a dot.",
  },
  message: {
    required: "Message is required",
    minLength: "Must be at least ten characters long.",
  },
};

export const emailRegex = /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$/;

export const labels = {
  name: "What is your name?",
  email: "What is your email?",
  message: "Your message",
};
