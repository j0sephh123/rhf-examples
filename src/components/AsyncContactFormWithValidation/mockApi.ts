import { ContactFormFields } from "./types";

export const mockSubmit = async (contactFormFields: ContactFormFields) => {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      if (
        contactFormFields.email &&
        contactFormFields.message &&
        contactFormFields.name
      ) {
        resolve(undefined);
      } else {
        reject("somethig went wrong");
      }
    }, 50);
  });
};
