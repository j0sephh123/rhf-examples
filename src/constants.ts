export const links = {
  "React Hook Form": "/form/react-hook-form",
  "Valtio Basics": "/valtio/basics",
  Copy: "/utility/copy",
};
