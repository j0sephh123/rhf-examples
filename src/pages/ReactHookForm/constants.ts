export const emailRegex = /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$/;

export const labels: ContactFormFields = {
  name: "What is your name?",
  email: "What is your email?",
  message: "Your message",
};

type ErrorOptions = Partial<{
  required: string;
  minLength: string;
  pattern: string;
}>;

export type ContactFormFields = {
  name: string;
  email: string;
  message: string;
};

export type ContactFormKeys = keyof ContactFormFields;

type FieldInfo = {
  label: string;
  errors: ErrorOptions;
};

export const contactFormInfo: Record<ContactFormKeys, FieldInfo> = {
  name: {
    label: "What is your name?",
    errors: {
      required: "Name is required",
      minLength: "Must be at least two characters long.",
    },
  },
  email: {
    label: "What is your email?",
    errors: {
      required: "Email is required",
      pattern: "Must include an '@' symbol and a dot.",
    },
  },
  message: {
    label: "Your message",
    errors: {
      required: "Message is required",
      minLength: "Must be at least ten characters long.",
    },
  },
};
