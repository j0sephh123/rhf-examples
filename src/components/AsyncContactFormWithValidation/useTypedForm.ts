import { RegisterOptions, useForm } from "react-hook-form";
import { ContactFormFields, ContactFormKeys } from "./types";
import { contactFormInfo, emailRegex } from "./constants";

type UseTypedFormOptions = {
  keys: ContactFormKeys[];
};

export default function useTypedForm({ keys }: UseTypedFormOptions) {
  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
    reset,
  } = useForm<ContactFormFields>({
    mode: "onSubmit",
    reValidateMode: "onSubmit",
  });

  const registerFields = keys.map((key) => {
    const fieldInfo = contactFormInfo[key];
    const registerOptions: RegisterOptions = {};
  
    if (fieldInfo.errors.required) {
      registerOptions.required = {
        value: true,
        message: fieldInfo.errors.required,
      };
    }
  
    if (fieldInfo.errors.minLength) {
      registerOptions.minLength = {
        value: key === "message" ? 10 : 2,
        message: fieldInfo.errors.minLength,
      };
    }
  
    if (key === "email" && fieldInfo.errors.pattern) {
      registerOptions.pattern = {
        value: emailRegex,
        message: fieldInfo.errors.pattern,
      };
    }
  
    return { key, registerOptions };
  });
  

  const registerRefs = registerFields.map(({ key, registerOptions }) =>
    register(key, registerOptions)
  );

  return {
    errors,
    handleSubmit,
    reset,
    isSubmitting,
    registerRefs, // This is an array of refs for each field
  };
}
