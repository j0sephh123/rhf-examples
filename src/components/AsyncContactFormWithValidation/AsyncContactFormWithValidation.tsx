import clsx from "clsx";
import { useForm } from "react-hook-form";
import { ContactFormFields } from "./types";
import { emailRegex, errorsMessages, labels } from "./constants";
import { mockSubmit } from "./mockApi";

export default function AsyncContactFormWithValidation() {
  const {
    register,
    handleSubmit,
    formState: { errors, isSubmitting },
    reset,
  } = useForm<ContactFormFields>({
    mode: "onSubmit",
    reValidateMode: "onSubmit",
  });

  const onSubmit = async (contactFormFields: ContactFormFields) => {
    await mockSubmit(contactFormFields);

    reset();
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className="form-control w-full max-w-xs">
        <label htmlFor="name" className="label">
          <span className="label-text">{labels.name}</span>
        </label>
        <input
          id="name"
          type="text"
          placeholder="Type here"
          className={clsx(
            "input input-bordered w-full max-w-xs",
            errors.name && "input-error"
          )}
          {...register("name", {
            required: {
              value: true,
              message: errorsMessages.name.required,
            },
            minLength: {
              message: errorsMessages.name.minLength,
              value: 2,
            },
          })}
        />
        {errors.name && (
          <p className="text-error text-sm mt-1">{errors.name.message}</p>
        )}
      </div>
      <div className="form-control w-full max-w-xs">
        <label htmlFor="email" className="label">
          <span className="label-text">{labels.email}</span>
        </label>
        <input
          id="email"
          type="text"
          placeholder="Type here"
          className={clsx(
            "input input-bordered w-full max-w-xs",
            errors.email && "input-error"
          )}
          {...register("email", {
            required: {
              value: true,
              message: errorsMessages.email.required,
            },
            pattern: {
              value: emailRegex,
              message: errorsMessages.email.pattern,
            },
          })}
        />
        {errors.email && (
          <p className="text-error text-sm mt-1">{errors.email.message}</p>
        )}
      </div>
      <div className="form-control">
        <label htmlFor="message" className="label">
          <span className="label-text">{labels.message}</span>
        </label>
        <textarea
          id="message"
          className={clsx(
            "textarea textarea-bordered h-24",
            errors.message && "textarea-error"
          )}
          placeholder="Bio"
          {...register("message", {
            required: {
              value: true,
              message: errorsMessages.message.required,
            },
            minLength: {
              value: 10,
              message: errorsMessages.message.minLength,
            },
          })}
        ></textarea>
        {errors.message && (
          <p className="text-error text-sm mt-1">{errors.message.message}</p>
        )}
      </div>
      <button
        type="submit"
        className={clsx(
          "mt-4 btn-neutral btn btn-block",
          isSubmitting && "btn-disabled"
        )}
      >
        {isSubmitting ? "Loading..." : "Submit"}
      </button>
    </form>
  );
}
