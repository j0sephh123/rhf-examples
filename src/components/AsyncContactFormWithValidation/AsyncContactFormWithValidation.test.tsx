import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi, describe, it } from "vitest";
import AsyncContactFormWithValidation from "./AsyncContactFormWithValidation";
import { errorsMessages, labels } from "./constants";
import * as mockApi from "./mockApi";

/**
  1. should render correct initial state
    a. button has "Submit" text and does not have "btn-disabled" class
    b. labels have default texts
    c. error paragraphs not rendering `errors.{ContactFormFields}`
    d. input and text area don't have error classes 
    e. fields are empty
  2. should correcty submit
    a. before
      i. enter correct values
      ii. click on submit button
    b. during
      i. button should be disabled, have loading class and loading text
      ii. api should get called with correct field values
    c. after
      i. button should have normal state
      ii. fields should be empty
  3. should correctly validate required fields
    a. before
      i. leave all fields empty, i.e do nothing
    b. after click submit
      i. validate that api is not called
      ii. validate that errors messages appear with correct text
    c. start typing once again
      i. enter correct values
      ii. re-use the logic from 2. to validate that it worked
  4. should correctly validate custom validations
    a. before clicking submit
      i. enter invalid fields for all 3, should not be empty
    b. after clicking submit
      i. check for all custom validation messages
      ii. re-use the logic from 2. to validate that it worked
 */

describe("AsyncContactFormWithValidation", () => {
  const getDOMElements = () => {
    const btn = screen.getByRole("button");
    const nameInput = screen.getByLabelText(labels.name);
    const emailInput = screen.getByLabelText(labels.email);
    const messageTextarea = screen.getByLabelText(labels.message);

    return {
      btn,
      nameInput,
      emailInput,
      messageTextarea,
    };
  };

  const assertButtonInitialState = (btn: HTMLElement) => {
    expect(btn).toHaveTextContent("Submit");
    expect(btn).not.toHaveClass("btn-disabled");
  };

  const assertButtonLoadingState = (btn: HTMLElement) => {
    expect(btn).toHaveClass("btn-disabled");
    expect(btn).toHaveTextContent("Loading...");
  };

  const assertFieldsAreEmpty = (fields: HTMLElement[]) => {
    fields.forEach((field) => expect(field).toHaveValue(""));
  };

  it("1. should render correct initial state", async () => {
    const { asFragment } = render(<AsyncContactFormWithValidation />);
    const { btn, emailInput, messageTextarea, nameInput } = getDOMElements();

    assertButtonInitialState(btn);

    expect(screen.getByText(labels.name)).toBeInTheDocument();
    expect(screen.getByText(labels.email)).toBeInTheDocument();
    expect(screen.getByText(labels.message)).toBeInTheDocument();

    expect(() => screen.getByText(errorsMessages.name.required)).toThrow();
    expect(() => screen.getByText(errorsMessages.name.minLength)).toThrow();
    expect(() => screen.getByText(errorsMessages.email.required)).toThrow();
    expect(() => screen.getByText(errorsMessages.email.pattern)).toThrow();
    expect(() => screen.getByText(errorsMessages.message.required)).toThrow();
    expect(() => screen.getByText(errorsMessages.message.minLength)).toThrow();

    expect(nameInput).not.toHaveClass("input-error");
    expect(emailInput).not.toHaveClass("input-error");
    expect(messageTextarea).not.toHaveClass("textarea-error");

    assertFieldsAreEmpty([nameInput, emailInput, messageTextarea]);

    expect(asFragment()).toMatchSnapshot();
  });
  it("2. should correcty submit", async () => {
    // TODO try again with vi.fakeTimers at some point
    const spy = vi.spyOn(mockApi, "mockSubmit");
    const user = userEvent.setup();

    const validName = "a valid name";
    const validEmail = "email@gmail.com";
    const validMessage = "a valid message";

    const { asFragment } = render(<AsyncContactFormWithValidation />);
    const { btn, emailInput, messageTextarea, nameInput } = getDOMElements();

    await user.type(nameInput, validName);
    expect(nameInput).toHaveValue(validName);
    await user.type(emailInput, validEmail);
    expect(emailInput).toHaveValue(validEmail);
    await user.type(messageTextarea, validMessage);
    expect(messageTextarea).toHaveValue(validMessage);

    await user.click(btn);

    expect(spy).toHaveBeenCalledWith({
      email: validEmail,
      message: validMessage,
      name: validName,
    });

    assertButtonLoadingState(btn);

    await waitFor(() => {
      assertButtonInitialState(btn);
    });

    assertFieldsAreEmpty([emailInput, nameInput, messageTextarea]);

    expect(asFragment()).toMatchSnapshot();
  });
  it("3. should correctly validate required fields", async () => {
    const user = userEvent.setup();
    const { asFragment } = render(<AsyncContactFormWithValidation />);

    const { btn, emailInput, messageTextarea, nameInput } = getDOMElements();

    await user.click(btn);

    expect(screen.getByText(errorsMessages.name.required)).toBeInTheDocument();
    expect(() => screen.getByText(errorsMessages.name.minLength)).toThrow();
    expect(screen.getByText(errorsMessages.email.required)).toBeInTheDocument();
    expect(() => screen.getByText(errorsMessages.email.pattern)).toThrow();
    expect(
      screen.getByText(errorsMessages.message.required)
    ).toBeInTheDocument();
    expect(() => screen.getByText(errorsMessages.message.minLength)).toThrow();

    expect(nameInput).toHaveClass("input-error");
    expect(emailInput).toHaveClass("input-error");
    expect(messageTextarea).toHaveClass("textarea-error");

    expect(asFragment()).toMatchSnapshot();
  });
  it("4. should correctly validate custom validations", async () => {
    const user = userEvent.setup();
    const { asFragment } = render(<AsyncContactFormWithValidation />);

    const { btn, emailInput, messageTextarea, nameInput } = getDOMElements();

    await user.type(nameInput, "1");
    await user.type(emailInput, "invalid mail");
    await user.type(messageTextarea, "short");

    await user.click(btn);

    expect(() => screen.getByText(errorsMessages.name.required)).toThrow();
    expect(screen.getByText(errorsMessages.name.minLength)).toBeInTheDocument();
    expect(() => screen.getByText(errorsMessages.email.required)).toThrow();
    expect(screen.getByText(errorsMessages.email.pattern)).toBeInTheDocument();
    expect(() => screen.getByText(errorsMessages.message.required)).toThrow();
    expect(
      screen.getByText(errorsMessages.message.minLength)
    ).toBeInTheDocument();

    expect(nameInput).toHaveClass("input-error");
    expect(emailInput).toHaveClass("input-error");
    expect(messageTextarea).toHaveClass("textarea-error");

    expect(asFragment()).toMatchSnapshot();
  });
});
