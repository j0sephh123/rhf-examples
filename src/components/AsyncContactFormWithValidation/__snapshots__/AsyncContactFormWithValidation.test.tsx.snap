// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AsyncContactFormWithValidation > 1. should render correct initial state 1`] = `
<DocumentFragment>
  <form>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
        for="name"
      >
        <span
          class="label-text"
        >
          What is your name?
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        id="name"
        name="name"
        placeholder="Type here"
        type="text"
      />
    </div>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
        for="email"
      >
        <span
          class="label-text"
        >
          What is your email?
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        id="email"
        name="email"
        placeholder="Type here"
        type="text"
      />
    </div>
    <div
      class="form-control"
    >
      <label
        class="label"
        for="message"
      >
        <span
          class="label-text"
        >
          Your message
        </span>
      </label>
      <textarea
        class="textarea textarea-bordered h-24"
        id="message"
        name="message"
        placeholder="Bio"
      />
    </div>
    <button
      class="mt-4 btn-neutral btn btn-block"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`AsyncContactFormWithValidation > 2. should correcty submit 1`] = `
<DocumentFragment>
  <form>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
        for="name"
      >
        <span
          class="label-text"
        >
          What is your name?
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        id="name"
        name="name"
        placeholder="Type here"
        type="text"
      />
    </div>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
        for="email"
      >
        <span
          class="label-text"
        >
          What is your email?
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs"
        id="email"
        name="email"
        placeholder="Type here"
        type="text"
      />
    </div>
    <div
      class="form-control"
    >
      <label
        class="label"
        for="message"
      >
        <span
          class="label-text"
        >
          Your message
        </span>
      </label>
      <textarea
        class="textarea textarea-bordered h-24"
        id="message"
        name="message"
        placeholder="Bio"
      />
    </div>
    <button
      class="mt-4 btn-neutral btn btn-block"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`AsyncContactFormWithValidation > 3. should correctly validate required fields 1`] = `
<DocumentFragment>
  <form>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
        for="name"
      >
        <span
          class="label-text"
        >
          What is your name?
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs input-error"
        id="name"
        name="name"
        placeholder="Type here"
        type="text"
      />
      <p
        class="text-error text-sm mt-1"
      >
        Name is required
      </p>
    </div>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
        for="email"
      >
        <span
          class="label-text"
        >
          What is your email?
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs input-error"
        id="email"
        name="email"
        placeholder="Type here"
        type="text"
      />
      <p
        class="text-error text-sm mt-1"
      >
        Email is required
      </p>
    </div>
    <div
      class="form-control"
    >
      <label
        class="label"
        for="message"
      >
        <span
          class="label-text"
        >
          Your message
        </span>
      </label>
      <textarea
        class="textarea textarea-bordered h-24 textarea-error"
        id="message"
        name="message"
        placeholder="Bio"
      />
      <p
        class="text-error text-sm mt-1"
      >
        Message is required
      </p>
    </div>
    <button
      class="mt-4 btn-neutral btn btn-block"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;

exports[`AsyncContactFormWithValidation > 4. should correctly validate custom validations 1`] = `
<DocumentFragment>
  <form>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
        for="name"
      >
        <span
          class="label-text"
        >
          What is your name?
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs input-error"
        id="name"
        name="name"
        placeholder="Type here"
        type="text"
      />
      <p
        class="text-error text-sm mt-1"
      >
        Must be at least two characters long.
      </p>
    </div>
    <div
      class="form-control w-full max-w-xs"
    >
      <label
        class="label"
        for="email"
      >
        <span
          class="label-text"
        >
          What is your email?
        </span>
      </label>
      <input
        class="input input-bordered w-full max-w-xs input-error"
        id="email"
        name="email"
        placeholder="Type here"
        type="text"
      />
      <p
        class="text-error text-sm mt-1"
      >
        Must include an '@' symbol and a dot.
      </p>
    </div>
    <div
      class="form-control"
    >
      <label
        class="label"
        for="message"
      >
        <span
          class="label-text"
        >
          Your message
        </span>
      </label>
      <textarea
        class="textarea textarea-bordered h-24 textarea-error"
        id="message"
        name="message"
        placeholder="Bio"
      />
      <p
        class="text-error text-sm mt-1"
      >
        Must be at least ten characters long.
      </p>
    </div>
    <button
      class="mt-4 btn-neutral btn btn-block"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
